### ISSUE TITLE
🔬  Triaging, Prioritizing, and Labeling Issues

### ISSUE THREAD

**Issue by joshrosso** [vmware-tanzu/community-edition#1829]
# 🔬  Triaging, Prioritizing, and Labeling Issues

This issue captures how the Tanzu Community Edition team labels and triages [issues](https://github.com/vmware-tanzu/community-edition/issues). It is meant to provide visibility into the inbound issues we receive.

## Milestones

[Milestones](https://github.com/vmware-tanzu/community-edition/milestones) are used to categorize work into a target release. Every triaged issue should have a milestone associated. Below are the milestone categories.

* `vX.XX.XX`: Release the work is targeting completion in.
    * See [v0.10.0](https://github.com/vmware-tanzu/community-edition/milestone/11) for an example.
* [Upcoming](https://github.com/vmware-tanzu/community-edition/milestone/14): Active or planned work that will be included in a future release. However, we don't know which one, yet.
* [Icebox](https://github.com/vmware-tanzu/community-edition/milestone/6): issues have no prioritization and their resolution is undetermined. Issues tagged with help wanted are not prioritized but we welcome contributions to bring them to resolution.

When planning a new release, we create a milestone and evaluate:

1. Should any work in `upcoming` be promoted here.
2. Should any work in `icebox` be promoted here.
3. Is there any net-new work we should create issues for and add here.

## Labeling

TCE operates on carefully chosen [labels](https://github.com/vmware-tanzu/tce/labels). Multiple teams at VMware use issues in this repository to track work and you may find a variety of labels. The **important** labels owned and used by the core engineering team have the following prefixes.

* **`kind/*`: The type of issue.**
    * `kind` powers our filtering ([see roadmap](https://github.com/vmware-tanzu/community-edition/issues/1293)) to understand what qualifies as a bug, feature, enhancement, or documentation.
* **`area/*`: The category of a `kind/issue`, `kind/bug`, or `kind/*`.**
    * `area/` is used for reporting and filtering issues so that we can understand which part of the project they pertain to.
* **`owner/*`: The team or project that would execute the work.**
    * `owner` is largely an internal detail, but enables us to acknowledge/assign a team or project ownership of work.
* **`triage/*`: The status of an issue.**
    * `triage` enables us to determine whether an issue has been looked at, responded to, and what follow-up is needed.
    * Issues absent of `triage/*`, have been triaged and are ready for work.

### `kind/*`

These labels are auto-assigned based on the [issue template](https://github.com/vmware-tanzu/community-edition/issues/new/choose) used. They may be re-assigned by TCE maintainers.

* `bug`: An issue that reports a defect in an existing `feature`.
* `feature`: An issue that requests a capability.
* `docs`: An issues that reports defects, improvements, or requests for new documentation.
* `test-release`: An issue that reports defects, improvements, or requests for new testing, automation, or release capabilities. These issues are **not** end-user facing.
* `feedback`: An issue that acts as a catch-all for issues. These issues are to be triaged and re-labeled with one of the above `kind/*`.

### `owner/*`

These labels are assigned by TCE maintainers as part of the triaging process.

* `core-eng`: An issue where work would be executed by the core TCE engineering team.
* `packages`: An issue where work would be executed by a package's maintainer(s).
* `docs`: An issue where work would be executed by VMware's documentation team.
* `release-eng`: An issue where work would be executed by VMware's release engineering team.
* `community`: An issue where work would be executed by VMware's community team.
* `framework`: An issue where work would be executed in [tanzu-framework](https://github.com/vmware-tanzu/tanzu-framework).
    * This issue type is often tracking an issue in [tanzu-framework](https://github.com/vmware-tanzu/tanzu-framework).
* `carvel`: An issue where work would be executed in a Carvel project such as `kapp-controller`, `ytt`, `imgpkg`, etc.
    * This issue type is often tracking an issue in a Carvel project.
* `upstream`: An issue where work would be executed in an upstream project that is not [tanzu-framework](https://github.com/vmware-tanzu/tanzu-framework) or Carvel. 
    * This issue type is often tracking an issue in an upstream project.

> We also accept external [contributions](https://github.com/vmware-tanzu/community-edition/blob/main/CONTRIBUTING.md)!

### `triage/*`

These labels are _sometimes_ assigned to issues that require additional follow-up. All net-new issues are marked with `need-triage`.

* `needs-triage`: An issue that still needs initial triage, response, and labeling from TCE maintainers.
* `blocked`: An issue where work cannot start or continue until something else is resolved.
* `needs-info`: An issue where more information is required from the reporter before work and start.
* `duplicate`: An issue where another issues covers the request.
    * These issues are typically closed after labeling.

### `package/*`

These labels are assigned by TCE maintainers on a best effort basis as part of the triaging process to help classify package specific issues.

### `provider/*`

These labels are assigned by TCE maintainers on a best effort basis as part of the triaging process to help classify provider specific issues.

* `aws`: Related to working with AWS.
* `azure`: Related to working with Azure.
* `docker`: Related to Docker and/or CAPD.
* `vSphere`: Related to working with vSphere.

### `area/*`

These labels act as categorical buckets for issues. Their primary use is filtering and reporting. The values may change over time.

## Triaging

Every new issue in TCE is marked as [triage/needs-triage](https://github.com/vmware-tanzu/tce/labels/triage%2Fneeds-triage). These issues have **not** been assigned a [milestone](#milestones).

When an issue is triaged, at minimum it is:

* Added to a milestone
* Assigned an `owner/*` label
* Assigned an `area/*` label
* Assigned an `owner/*` label
* Assigned a `kind/*` label

![image](https://user-images.githubusercontent.com/6200057/133945295-9060f073-e5e2-4838-987d-e5486830149f.png)

As a project with multiple dependencies, some issues may required opening an ancillary issue in a dependency's repository. When this takes place, the original TCE issue will becoming a tracking issue for the dependency's issue.

### Projects

Projects are a 1:1 reflection of a [milestone](#milestones). They offer engineering a means to understand what is:

* **Todo:** Planned but not actively worked on
* **In Progress:** Being worked on
  * These issues **must** have an associated PR linked to them

    ![image](https://user-images.githubusercontent.com/6200057/133945319-f4f75bb4-8b01-49a6-9a17-3a617d889ece.png)

* **In Review:** PR is open
  * These PRs must be **not** marked as drafts
* **Completed:** PR has merged and issue is closed

Pull requests should **not** be added to projects, instead they are linked to issues in the project.

## Common Questions and Answers

* **Q:** How do I know what is slated for a release?
  * **A:** There are multiple ways:
        - View the [roadmap](https://github.com/vmware-tanzu/tce/issues/1293).
        - View the corresponding milestone ([for example, v0.8.0](https://github.com/vmware-tanzu/community-edition/milestone/10)).
* **Q:** How do I view the progress on an in-flight release?
  * **A:** View the corresponding project for that release/milestone ([for example, v0.8.0](https://github.com/vmware-tanzu/tce/projects/10)).
* **Q:** How do I view bugs specifically being targeting for a milestone?
  * **A:** Create an issue filter for `kind/bug` on the given milestone ([for example, bugs fixed in v0.8.0](https://github.com/vmware-tanzu/tce/issues?q=label%3Akind%2Fbug+milestone%3Av0.8.0+)).
* **Q**: How do I view issues relating to documentation for a milestone?
  * **A:** Create an issue filter for `area/docs` on the given milestone ([for example, docs issues in v0.8.0](https://github.com/vmware-tanzu/community-edition/issues?q=label%3Aarea%2Fdocs+milestone%3Av0.8.0+label%3Akind%2Fdocs)).
